crates/xilem_masonry/src/sequence.rs

```diff
@@ -0,0 +1,221 @@
+        message: Box<dyn std::any::Any>,
+        app_state: &mut State,
+    ) -> MessageResult<Action>;
+
+    /// Returns the current amount of widgets built by this sequence.
+    fn count(&self) -> usize;
+}
+
+/// Workaround for trait ambiguity
+///
+/// These need to be public for type inference
+#[doc(hidden)]
+pub struct WasAView;
+#[doc(hidden)]
+/// See [`WasAView`]
+pub struct WasASequence;
+
+impl<State, Action, View: MasonryView<State, Action>> ViewSequence<State, Action, WasAView>
+    for View
+{
+    fn build(&self, cx: &mut ViewCx, elements: &mut dyn ElementSplice) {
+        let element = self.build(cx);
+        elements.push(element.boxed());
+    }
+
+    fn rebuild(
+        &self,
+        cx: &mut ViewCx,
+        prev: &Self,
+        elements: &mut dyn ElementSplice,
+    ) -> ChangeFlags {
+        let mut element = elements.mutate();
+        let downcast = element.downcast::<View::Element>();
+
+        if let Some(element) = downcast {
+            self.rebuild(cx, prev, element)
+        } else {
+            unreachable!("Tree structure tracking got wrong element type")
+        }
+    }
+
+    fn message(
+        &self,
+        id_path: &[ViewId],
+        message: Box<dyn std::any::Any>,
+        app_state: &mut State,
+    ) -> MessageResult<Action> {
+        self.message(id_path, message, app_state)
+    }
+
+    fn count(&self) -> usize {
+        1
+    }
+}
+
+impl<State, Action, Marker, VT: ViewSequence<State, Action, Marker>>
+    ViewSequence<State, Action, (WasASequence, Marker)> for Option<VT>
+{
+    fn build(&self, cx: &mut ViewCx, elements: &mut dyn ElementSplice) {
+        match self {
+            Some(this) => {
+                this.build(cx, elements);
+            }
+            None => (),
+        }
+    }
+
+    fn rebuild(
+        &self,
+        cx: &mut ViewCx,
+        prev: &Self,
+        elements: &mut dyn ElementSplice,
+    ) -> ChangeFlags {
+        match (self, prev) {
+            (Some(this), Some(prev)) => this.rebuild(cx, prev, elements),
+            (None, Some(prev)) => {
+                let count = prev.count();
+                elements.delete(count);
+
+                ChangeFlags::CHANGED
+            }
+            (Some(this), None) => {
+                this.build(cx, elements);
+                ChangeFlags::CHANGED
+            }
+            (None, None) => ChangeFlags::UNCHANGED,
+        }
+    }
+
+    fn message(
+        &self,
+        id_path: &[ViewId],
+        message: Box<dyn std::any::Any>,
+        app_state: &mut State,
+    ) -> MessageResult<Action> {
+        if let Some(this) = self {
+            this.message(id_path, message, app_state)
+        } else {
+            MessageResult::Stale(message)
+        }
+    }
+
+    fn count(&self) -> usize {
+        match self {
+            Some(this) => this.count(),
+            None => 0,
+        }
+    }
+}
+
+// TODO: We use raw indexing for this value. What would make it invalid?
```

Comment:
The design of `ViewSequence` here is subtly different to previous versions of Xilem.

Currently, each non-trivial `ViewSequence`s add a level to the ViewId path, and they have complete control of their own `ViewId`.

This is a quite different, much sparser model of view ids